File: web/src/lib/schemas/alert.ts
import { z } from "zod";

// The quant alert bus serializes Decimal value/threshold as JSON strings (same no-float-in-money
// contract as signals); coerce to number for display. Both are null for thresholdless alerts.
const money = z.coerce.number().nullable();

export const AlertKindSchema = z.enum(["ws_drop", "drawdown_breach", "calibration_drift"]);
export type AlertKind = z.infer<typeof AlertKindSchema>;

export const AlertSeveritySchema = z.enum(["info", "warning", "error"]);
export type AlertSeverity = z.infer<typeof AlertSeveritySchema>;

export const AlertSchema = z.object({
  kind: AlertKindSchema,
  severity: AlertSeveritySchema,
  title: z.string(),
  detail: z.string(),
  value: money,
  threshold: money,
  time: z.string(), // ISO-8601 (UTC)
});
export type Alert = z.infer<typeof AlertSchema>;

File: quant/app/models/alert.py
"""Alert model — one operational alert (frozen, Decimal-native).

Produced by the pure predicates in ``app.observability.alerts``, published to the Redis
``edge:alerts`` channel and captured to Sentry by the alert bus, then surfaced as a dashboard
toast. Pydantic renders ``Decimal`` as a JSON **string** — the same Decimal->string wire
contract as ``AdvisedSignal``, so the web Zod boundary coerces it the same way.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict

AlertKind = Literal["ws_drop", "drawdown_breach", "calibration_drift"]
AlertSeverity = Literal["info", "warning", "error"]


class Alert(BaseModel):
    model_config = ConfigDict(frozen=True)

    kind: AlertKind
    severity: AlertSeverity
    title: str
    detail: str
    value: Decimal | None  # observed metric (drawdown fraction, drift gap, drop count)
    threshold: Decimal | None  # the breached threshold (None for thresholdless alerts)
    time: datetime  # UTC, injected by the caller (test seam)
